### PR TITLE
Fixed sitemap regressions

### PR DESCRIPTION
**Pull request by rshbhgrg**
refs #10640 and [Forum](https://forum.ghost.org/t/images-not-included-in-the-sitemap/6692/1)

- Updated resources config to include certain fields needed for correct sitemap generation
--------------
### Context

We have seen 2 main regressions that have popped up for sitemaps - 
- Incorrect last modified Date - #10640 
- Incorrect image count - [Forum](https://forum.ghost.org/t/images-not-included-in-the-sitemap/6692/1)

The value calculation of both last modified date and image count are part of sitemap generation in [base generator](https://github.com/TryGhost/Ghost/blob/master/core/server/data/xml/sitemap/base-generator.js), which applies to all sitemap resources. These values are calculated based on values of certain fields for the resource, which in case of date and image are calculated here - 
- For Date: https://github.com/TryGhost/Ghost/blob/master/core/server/data/xml/sitemap/base-generator.js#L68
- For Images: https://github.com/TryGhost/Ghost/blob/master/core/server/data/xml/sitemap/base-generator.js#L100

### Problem

Digging into data available for individual resource, the root cause for both issues is unavailability of some or all of date/image fields on the resources(tags, posts, pages, authors) during sitemap generation. As a result, since fields like `created_at`, `updated_at`, `feature_image` etc are not available on resource, the resulting values calculated are incorrect or incomplete on sitemap.

Sitemap base generator gets resources data generated from [Resources](https://github.com/TryGhost/Ghost/blob/master/core/server/services/url/Resources.js) file. The Resources file uses the config as [here](https://github.com/TryGhost/Ghost/blob/master/core/server/services/url/Resources.js#L49) to determine allowed fields on the resource data. The [config](https://github.com/TryGhost/Ghost/blob/master/core/server/services/url/configs/v2.js) itself excludes a lot of fields from different resources, which eventually leads to incorrect sitemap calculation.

### Possible solutions

🤔 This one seems to be a bit tricky on quick look to solve, without affecting either the global url config file or Resources config. While changing the config file is a quick fix for sitemap issue(tested locally), it needs to be seen if it has any other unwanted side-affects.(cc @gargol)

### CODE REVIEW

**Comment by gargol** [comment]
@rishabhgrg good analysis of the problem :+1: Have double checked on both of them and here are some additional findings.
1. RE incorrect last modified date. The fix proposed here is good but I think it's just a shim of the real problem which I suspect was introduced [here](https://github.com/TryGhost/Ghost/commit/ef98c65040ec2c901459bda20c06c8c287bab9b8#diff-b67159d3704769e52d84b34b0499b59dL88). When doing a cleanup of fields returned by Content API. the `created_at` & `updated_at` were removed but they actually should've been left alone because they are needed for correct sitemap generation. We don't expose these resources to the outside and only use them internally to build URLs and sitemaps.
2. RE incorrect image count. Your suggestion to correct config files (remove `feature_image` from excludes in post/page resources) is on spot :+1:  The reason we have excludes in resource configs is to have as quick queries as possible and fetch only data needed to later generate urls/sitemaps. The speed is important because this fetching is happening during Ghost startup, so the more we have in the query the longer it takes. `feature_image` should've not appeared on that list in the first place.

Additionally, the image count is a bit artificial at the moment. It only takes into account few fields from each resource respectively `cover_image`, `profile_image`, `feature_image`. These fields don't represent all images that are present in the post content for example. But this is probably a topic for when we rejig our sitemaps in the future :smiley:  

Summing up, think your proposed solution correcting resource config files  is the best way to tackle both issues :+1:

**Comment by rishabhgrg** [comment]
@gargol Thanks, this is very useful, exactly the context I was looking for. 🤗 My main concern of exposing the additional fields was what areas they affect, it makes sense to expose essential fields needed for sitemap generation with above context. 

💯 agree on the artificial image count in the sitemaps atm, had same feeling while going through and this means it will mostly just show `0` or `1` as image count for any post, but not the true number. The same behaviour also leaks in gatsby sitemap generator, and as you said, its def part of bigger rejig of sitemaps, will add it to separate issue.

I'll update the PR with resource config fix for a quick review. 😄

File: core/server/data/xml/sitemap/base-generator.js [hunk 1]
```diff
@@ -82,7 +88,7 @@ class BaseSiteMapGenerator {
         node = {
             url: [
                 {loc: url},
-                {lastmod: moment(this.getLastModifiedForDatum(datum)).toISOString()}
+                {lastmod: moment(this.getLastModifiedForDatum(datum), moment.ISO_8601).toISOString()}
```
Comment (gargol) [inline core/server/data/xml/sitemap/base-generator.js]: Don't think `moment.ISO_8601` is necessary here as `toISOString()` [already applies ](https://momentjs.com/docs/#/displaying/as-iso-string/) ISO_8601 formatting automatically

File: core/server/data/xml/sitemap/base-generator.js [hunk 2]
```diff
@@ -65,7 +65,13 @@ class BaseSiteMapGenerator {
     }
 
     getLastModifiedForDatum(datum) {
-        return datum.updated_at || datum.published_at || datum.created_at;
+        if (datum.updated_at || datum.published_at || datum.created_at) {
+            const modifiedDate = datum.updated_at || datum.published_at || datum.created_at;
+
+            return moment(new Date(modifiedDate));
```
Comment (gargol) [inline core/server/data/xml/sitemap/base-generator.js]: have checked the usages in the codebase, think it's unnecessary here and below to wrap the date in additional `new Date()` constructor. [See](https://momentjs.com/docs/#/parsing/now/) explanation :wink:

File: core/server/services/url/configs/v2.js [hunk 3]
```diff
@@ -30,7 +30,6 @@ module.exports = [
                 'twitter_title',
                 'twitter_description',
                 'custom_template',
-                'feature_image',
```
Comment (gargol) [inline core/server/services/url/configs/v2.js]: The same change should be applied for `v0.1` schema :nerd_face:

**Review by naz — CHANGES_REQUESTED** [review]
Just a couple of minor suggestions and v01 correction :+1:

File: core/server/data/xml/sitemap/base-generator.js (hunk 1 above)
Comment (rishabhgrg) [reply core/server/data/xml/sitemap/base-generator.js]: Fair enough, seems it can be removed safely. Added it based on its usage in gatsby plugin [here](https://github.com/TryGhost/gatsby-plugin-advanced-sitemap/blob/master/src/BaseSiteMapGenerator.js#L91) 😄

**Comment by rishabhgrg** [comment]
@gargol Thanks for the quick review, updated on all counts! 😄

**Review by naz — APPROVED** [review]
